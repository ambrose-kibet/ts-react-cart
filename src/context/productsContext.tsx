import {
  ReactElement,
  createContext,
  useState,
  useContext,
  useEffect,
  useCallback,
} from 'react';

import axios from 'axios';
export type productType = {
  id: number;
  title: string;
  price: number;
  img: string;
  amount: number;
};
type contextType = {
  children?: ReactElement;
};
type useProductsContextType = {
  products: productType[];
  isLoading: boolean;
};
const url = 'https://course-api.com/react-useReducer-cart-project';
export const ProductsContext = createContext<useProductsContextType>({
  products: [],
  isLoading: false,
});
const ProductsProvider = ({ children }: contextType) => {
  const [products, setProducts] = useState<useProductsContextType>({
    products: [],
    isLoading: false,
  });
  const fetchProducts = useCallback(async () => {
    try {
      setProducts({ ...products, isLoading: true });
      const { data } = await axios(url);
      setProducts({ products: data, isLoading: false });
    } catch (error) {
      console.log(error);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  useEffect(() => {
    fetchProducts();
  }, [fetchProducts]);

  return (
    <ProductsContext.Provider value={products}>
      {children}
    </ProductsContext.Provider>
  );
};
export default ProductsProvider;

// eslint-disable-next-line react-refresh/only-export-components
export const useProductsContext = (): useProductsContextType => {
  return useContext(ProductsContext);
};
