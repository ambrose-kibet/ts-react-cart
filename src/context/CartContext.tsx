import {
  createContext,
  ReactElement,
  useReducer,
  useContext,
  useEffect,
} from 'react';
import { productType } from './productsContext';
import { CART_REDUCER_ACTIONS } from '../actions';
import reducer from '../reducers/cartReducer';

export type cart = {
  cartItems: productType[];
  total: number;
  amount: number;
};

type ChildrenType = {
  children?: ReactElement | ReactElement[];
};
export type payloadType = {
  id: number;
  product?: productType;
  type?: string;
};

export type ActionType = {
  type: string;
  payload?: payloadType;
};
type contextActions = {
  addToCart: ({ id, product }: payloadType) => void;
  removeFromCart: ({ id }: payloadType) => void;
  clearCart: () => void;
  toggleAmount: ({ id, type }: payloadType) => void;
};
const initialState: cart = {
  cartItems: [],
  total: 0,
  amount: 0,
};

const CartContext = createContext<cart & contextActions>({
  ...initialState,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  addToCart: (payload: payloadType) => {},
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  removeFromCart: ({ id }: payloadType) => {},
  clearCart: () => {},
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  toggleAmount: (payload: payloadType) => {},
});

const CartProvider = ({ children }: ChildrenType) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const addToCart = ({ id, product }: payloadType): void => {
    dispatch({
      type: CART_REDUCER_ACTIONS.ADD_TO_CART,
      payload: { id, product },
    });
  };
  const removeFromCart = ({ id }: payloadType) => {
    dispatch({ type: CART_REDUCER_ACTIONS.REMOVE_FROM_CART, payload: { id } });
  };
  const clearCart = () => {
    dispatch({ type: CART_REDUCER_ACTIONS.CLEAR_CART });
  };
  const toggleAmount = ({ id, type }: payloadType) => {
    dispatch({
      type: CART_REDUCER_ACTIONS.TOGGLE_AMOUNT,
      payload: { id, type },
    });
  };
  const getTotals = () => {
    dispatch({ type: CART_REDUCER_ACTIONS.GET_TOTALS });
  };
  useEffect(() => {
    getTotals();
  }, [state.cartItems]);

  return (
    <CartContext.Provider
      value={{ ...state, addToCart, removeFromCart, clearCart, toggleAmount }}
    >
      {children}
    </CartContext.Provider>
  );
};
export default CartProvider;
export const useCartContext = (): cart & contextActions =>
  useContext(CartContext);
