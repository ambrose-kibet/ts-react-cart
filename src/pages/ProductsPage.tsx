import SingleProduct from '../components/SingleProduct';
import { useProductsContext } from '../context/productsContext';

const ProductsPage = () => {
  const { products, isLoading } = useProductsContext();
  console.log(products);
  if (isLoading) {
    return <h1 style={{ textAlign: 'center' }}> Loading...</h1>;
  }
  return (
    <section className="products-section">
      {products.map((product) => (
        <SingleProduct product={product} key={product.id} />
      ))}
    </section>
  );
};
export default ProductsPage;
