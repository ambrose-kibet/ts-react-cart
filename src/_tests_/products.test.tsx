import { screen, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { describe, it, expect } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import { ProductsContext } from '../context/productsContext';
import ProductsPage from '../pages/ProductsPage';

describe('Render products page', () => {
  const productList = [
    {
      id: 2,
      title: 'google pixel ',
      price: 499.99,
      img: 'https://res.cloudinary.com/diqqf3eq2/image/upload/v1583371867/phone-1_gvesln.png',
      amount: 1,
    },
    {
      id: 3,
      title: 'Xiaomi Redmi Note 2',
      price: 699.99,
      img: 'https://res.cloudinary.com/diqqf3eq2/image/upload/v1583368224/phone-3_h2s6fo.png',
      amount: 1,
    },
  ];

  it('should render correctly', () => {
    const { container } = render(
      <ProductsContext.Provider
        value={{ products: productList, isLoading: false }}
      >
        <MemoryRouter>
          <ProductsPage />
        </MemoryRouter>
      </ProductsContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
  it('should have all products on the screen', () => {
    render(
      <ProductsContext.Provider
        value={{ products: productList, isLoading: false }}
      >
        <MemoryRouter>
          <ProductsPage />
        </MemoryRouter>
      </ProductsContext.Provider>
    );
    const getByTextWithSpace = (text: string) => {
      return screen.getByText((content) => {
        return content.trim().includes(text.trim());
      });
    };
    productList.forEach((product) => {
      expect(getByTextWithSpace(product.title)).toBeInTheDocument();
      expect(getByTextWithSpace(`${product.price}`)).toBeInTheDocument();
    });
  });
  it('should render product images correctly', () => {
    render(
      <ProductsContext.Provider
        value={{ products: productList, isLoading: false }}
      >
        <MemoryRouter>
          <ProductsPage />
        </MemoryRouter>
      </ProductsContext.Provider>
    );
    const productImageElements: HTMLImageElement[] =
      screen.getAllByTestId('product-image');
    expect(productImageElements).toHaveLength(productList.length);

    productImageElements.forEach((element) => {
      expect(element.src).toMatch(/\.png/);
    });
  });
});
