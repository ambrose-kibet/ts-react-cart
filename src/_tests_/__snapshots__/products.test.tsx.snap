// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render products page > should render correctly 1`] = `
<div>
  <section
    class="products-section"
  >
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="google pixel "
          data-testid="product-image"
          src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1583371867/phone-1_gvesln.png"
          width="100%"
        />
      </div>
      <div
        class="card-body"
      >
        <h5>
          google pixel 
        </h5>
        <h5>
           $
          499.99
        </h5>
      </div>
      <div
        class="card-footer"
      >
        <a
          class="btn"
          href="/cart"
        >
          Add to cart
        </a>
      </div>
    </article>
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="Xiaomi Redmi Note 2"
          data-testid="product-image"
          src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1583368224/phone-3_h2s6fo.png"
          width="100%"
        />
      </div>
      <div
        class="card-body"
      >
        <h5>
          Xiaomi Redmi Note 2
        </h5>
        <h5>
           $
          699.99
        </h5>
      </div>
      <div
        class="card-footer"
      >
        <a
          class="btn"
          href="/cart"
        >
          Add to cart
        </a>
      </div>
    </article>
  </section>
</div>
`;
